zk/legacy_executor_verifier/legacy_executor_verifier.go

```diff
@@ -197,45 +234,59 @@ func (v *LegacyExecutorVerifier) AddRequestUnsafe(ctx context.Context, tx kv.RwT
 		// 	atomic.StoreInt32(&counter, 1)
 		// }
 
-		response := &VerifierResponse{
+		return NewVerifierBundle(request, &VerifierResponse{
 			BatchNumber:      request.BatchNumber,
 			Valid:            ok,
 			Witness:          witness,
 			ExecutorResponse: executorResponse,
 			Error:            executorErr,
-		}
-
-		return response, nil
+		}), nil
 	})
 
 	// add batch to the list of batches we've added
 	v.addedBatches[request.BatchNumber] = struct{}{}
 
 	// add the promise to the list of promises
 	v.promises = append(v.promises, promise)
-	return promise, nil
+	return promise
 }
 
-func (v *LegacyExecutorVerifier) ConsumeResultsUnsafe(tx kv.RwTx) ([]*VerifierResponse, error) {
+// Unsafe is not thread-safe so it MUST be invoked only from a single thread
+func (v *LegacyExecutorVerifier) ProcessResultsUnsafe(tx kv.RwTx) ([]*VerifierResponse, error) {
 	hdb := hermez_db.NewHermezDbReader(tx)
 
 	results := make([]*VerifierResponse, 0, len(v.promises))
-	for _, promise := range v.promises {
-		result, err := promise.GetNonBlocking()
-		if result == nil && err == nil {
+	for i := 0; i < len(v.promises); i++ {
+		verifierBundle, err := v.promises[i].TryGet()
+		if verifierBundle == nil && err == nil {
 			break
 		}
+
+		verifierResponse := verifierBundle.response
 		if err != nil {
-			log.Error("error getting verifier result", "err", err)
+			log.Error("error on our end while preparing the verification request, re-queueing the task", "err", err)
+			// this is an error on our end, so just re-create the promise at exact position where it was
+			if verifierBundle.request.isOverdue() {
+				return nil, fmt.Errorf("error: batch %d couldn't be processed in 30 minutes", verifierBundle.request.BatchNumber)
+			}
+
+			v.promises[i] = NewPromise[*VerifierBundle](v.promises[i].task)
+			break
 		}
-		if result != nil {
-			if err = v.writeBatchToStream(result, hdb, tx); err != nil {
+
+		if verifierResponse.Valid {
+			if err = v.WriteBatchToStream(verifierResponse.BatchNumber, hdb, tx); err != nil {
 				log.Error("error getting verifier result", "err", err)
 			}
 		}
 
-		results = append(results, result)
-		delete(v.addedBatches, result.BatchNumber)
+		results = append(results, verifierResponse)
+		delete(v.addedBatches, verifierResponse.BatchNumber)
+
+		// no point to process any further responses if we've found an invalid one
+		if !verifierResponse.Valid {
+			break
```

Comment:
lets put a comment here about limbo and maybe log to the console that something really bad has happened

Reply:
This is just a helper function that is used during verification stage. The actual log is on line 125 at stage_sequencer_executor_verify.go.  Do you think we should duplicate the logs?

I renamed the function to explicitly state that the results are processed sequentially.

Reply:
No problem if the logging happens upstream, I didn't look that far.